velox/docs/develop/hash-table.rst

```diff
@@ -0,0 +1,117 @@
+==========
+Hash table
+==========
+
+
+Layout
+------
+
+HashTable is implemented as an array of buckets. It is a linear data structure.
+Each bucket uses 128 bytes (2 * 64 = 2 cache lines) and contains 16 slots.
+Each HashTable entry occupies one slot. HashTable’s capacity is the total number of slots: total
+number of buckets * 16. HashTable’s capacity is always a power of 2.
+
+Each bucket contains 16 slots. Each slot consists of 2 pieces: a tag (7 bits) and a pointer (6 bytes).
+There are a total of 16 tags and 16 pointers in a bucket. These are stored tags first, followed by
+pointers. Each tag occupies 1 byte (only 7 bits are used). 16 tags occupy 16 bytes. Each pointer
+occupies 6 bytes. 16 pointers occupy 48 bytes. There are 16 bytes left unused at the end of the bucket.
+These are referred to as padding.
+
+.. image:: images/hash-table-layout.png
+    :align: center
+
+A hash table is never full. There are always some empty slots. Velox allows HashTable to fill up to ⅞ of
+capacity before resizing.
+On resize HashTable’s capacity doubles.
+
+Individual buckets may be completely empty, partially filled or full. Buckets are filled left to right.
+If a bucket is partially full, then first N tags and N pointers are filled and the rest are free (N < 16).
+
+
+Inserting an entry
+------------------
+
+To insert a new entry we need to figure out which slot to put it in.
+A slot is identified by bucket and offset within the bucket. First, we compute a hash of the entry.
+Then, we compute a tag and a bucket number from the hash.
+
+We use 7 bits of the hash for the tag: bits 32-38 inclusive. We use N bits of the hash for the bucket
+starting from bit 8.
+
+Given hash number
+
+01011011 11001010 01111100 0\ **1101001** 10110111 1\ **0010100 11111000 1**\ 1001110
```

Comment:
I wonder if this renders nicely. In Google doc I used colors to mark the bits. Wondering if this is possible here.

<img width="729" alt="Screenshot 2024-01-02 at 10 03 53 AM" src="https://github.com/facebookincubator/velox/assets/27965151/3943a7a9-be1e-4d95-b866-5976f67c5075">